src/effects/VST3/VST3Instance.cpp

```diff
@@ -0,0 +1,194 @@
+#include "VST3Instance.h"
+
+#include <public.sdk/source/vst/utility/uid.h>
+
+#include "AudacityException.h"
+#include "VST3Wrapper.h"
+#include "ConfigInterface.h"
+#include "VST3Utils.h"
+
+namespace
+{
+   unsigned CountChannels(Steinberg::Vst::IComponent* component,
+   const Steinberg::Vst::MediaTypes mediaType, 
+   const Steinberg::Vst::BusDirection busDirection,
+   const Steinberg::Vst::BusType busType)
+   {
+      using namespace Steinberg;
+
+      unsigned channelsCount{0};
+
+      const auto busCount = component->getBusCount(mediaType, busDirection);
+      for(auto i = 0; i < busCount; ++i)
+      {
+         Vst::BusInfo busInfo;
+         if(component->getBusInfo(mediaType, busDirection, i, busInfo) == kResultOk)
+         {
+            if(busInfo.busType == busType)
+               channelsCount += busInfo.channelCount;
+         }
+      }
+      return channelsCount;
+   }
+}
+
+VST3Instance::VST3Instance(const PerTrackEffect& effect, VST3::Hosting::Module& module, VST3::UID effectUID)
+   : Instance(effect), mEffectUID(effectUID)
+{
+   ReloadUserOptions();
+   mWrapper = std::make_unique<VST3Wrapper>(module, mEffectUID);
+}
+
+VST3Instance::~VST3Instance() = default;
+
+size_t VST3Instance::GetTailSize() const
+{
+   return Instance::GetTailSize();
+}
+
+bool VST3Instance::Init()
+{
+   return Instance::Init();
+}
+
+bool VST3Instance::RealtimeAddProcessor(EffectSettings& settings, unsigned, float sampleRate)
+{
+   return true;
+}
+
+bool VST3Instance::RealtimeFinalize(EffectSettings& settings) noexcept
+{
+   mWrapper->Finalize();
+   return true;
+}
+
+bool VST3Instance::RealtimeInitialize(EffectSettings& settings, double sampleRate)
+{
+   if(mWrapper->Initialize(settings, sampleRate, Steinberg::Vst::kRealtime, mProcessingBlockSize))
+   {
+      mInitialDelay = mWrapper->GetLatencySamples();
+      return true;
+   }
+   return false;
+}
+
+size_t VST3Instance::RealtimeProcess(size_t group, EffectSettings& settings, const float* const* inBuf,
+   float* const* outBuf, size_t numSamples)
+{
+   if(group == 0)
+      return mWrapper->Process(inBuf, outBuf, numSamples);
+   return 0;
+}
+
+bool VST3Instance::RealtimeProcessEnd(EffectSettings& settings) noexcept
+{
+   return true;
+}
+
+bool VST3Instance::RealtimeProcessStart(EffectSettings& settings)
+{
+   mWrapper->ConsumeChanges(settings);
```

Comment:
This was the added `ConsumeChanges` call I was expecting to see -- good, but again, I don't understand why it is right to remove the other calls, and it also needs rewrite to avoid deallocations.

Reply:
1. There is only one instance of VST3Wrapper which owned by VST3Instance (no per-track sub effect initialization)
2. There is only one point where actual parameter change transfer before processing could occur - EffectInstance::RealtimeProcessStart
3. In case of destructive processing instance state updated via IComponent::setState with settings stored by VST3UIValidator::ValidateUI